runtime/p2p.ss

```diff
@@ -0,0 +1,191 @@
+  (bio-force-output (stream-out s)))
+
+;;~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Pubsub methods for Retreiving, Creating, and Adding data for ethaddr->multiaddr~~~~~~~~~~~~~~~~~
+
+
+;;connect-to-multiaddr-pubsub
+;; pubsub-node: the multiaddr of a pubsub node to connect to
+;; c: a libp2p client object representing you
+;; connects to the provided pubsub node and subscribes to the "chat" pubsub
+;; also spawns a reader that will respond to identify requests
+;;
+;; returns (values pubsubchatchannel cancelprocforpubsubchannel readerthread)
+(def (connect-to-multiaddr-pubsub pubsub-node: pubsub-node c: c my-0xaddr: my-0xaddr)
+  (libp2p-connect c (string->peer-info pubsub-node))
+  (let*-values (((sub cancel) (pubsub-subscribe c "chat"))
+                (reader (spawn subscription-reader sub c my-0xaddr)))
+    (values sub cancel reader)))
+
+;; subscription-reader
+;; sub : the pubsub channel
+;; c : the libp2p client object
+;; my-0xaddr : the blockchain address of you
+;; will read the pubsub channel and when requested to identify itself, it will broadcast the client's identity
+;;
+;; TODO: Allow for specific identity requests
+(def (subscription-reader sub c my-0xaddr)
+ (let ((self (libp2p-identify c)))
+   (for (m sub)
+
+     ;;if message says "IDENTIFY" then publish your 0xaddr, else do nothing
+     (if (string=? "IDENTIFY" (bytes->string (vector-ref m 1)))
+       (pubsub-publish c "chat" (string->bytes my-0xaddr))))))
+
+;; get-peerID-from-pubsub
+;;
+;; sub : the pubsub chat channel
+;; c : your libp2p client
+;; peer0x : the peer you are waiting for to come online
+;;
+;; ask everyone in pubsub channel to IDENTIFY, then read messages until you see one matching peer0x
+;; block until peerID is found
+;; TODO: verify that peer0x comes from correct sender
+;; returns the peer-ID related to peer0x
+(def (get-peerID-from-pubsub sub
+                             c
+                             peer0x
+                             timeout)
+  (pubsub-publish c "chat" (string->bytes "IDENTIFY"))
+
+  (let lp ()
+    (let ((m (channel-get sub 5)))
+      (if m
+        ;;if the message matches the peer0x you are requesting, return the peerID, else move on TODO: VERIFY FROM CORRECT SENDER AS WELL
+        (if (string=? peer0x (bytes->string (vector-ref m 1)))
+          (ID->string (vector-ref m 0)) ;;FIXME: Does this return the id??
+          (lp))
+
+        (if (and timeout (> timeout 0))
+          (begin
+            (displayln "Failed to find peer ... Trying again ...")
+            (set! timeout (- timeout 1))
+            (pubsub-publish c "chat" (string->bytes "IDENTIFY"))
+            (lp))
+          (error "Timeout when getting peerID from pubsub"))))))
+
+
+
+
+;;~~~~~~~~~~~~~~~~~~~~~~~~~~ Glow Bootstrap Node Methods for testing ~~~~~~~~~~~~~~~~~~~
+
+
+
+;; do-bootstrap
+;; host-addresses: a string of the ip address in multiaddr format to start the bootstrap node at
+;;
+;; returns: a libp2p client object of the bootstrap node
```

Comment:
This seems incorrect, since we're returning two values?

Reply:
Good Catch! Completely forgot to change that. Fixing now